### ISSUE TITLE
如何获取FARTrack-Tiny/FARTrack-Nano/FARTrack-Pico的预训练模型？

### ISSUE THREAD

**Issue by canquanshen** [MIV-XJTU/FARTrack#1]
非常感谢你开源源代码，请问如何获取FARTrack-Tiny/FARTrack-Nano/FARTrack-Pico的预训练模型？

**Comment by AlexDotHam** [comment]
直接截断tiny的权重至6层和10层即可，输出时也直接使用对应层的追踪结果

**Comment by canquanshen** [comment]
非常感谢你的回复，我看了开源的代码，没有FARTrack-Tiny/FARTrack-Nano/FARTrack-Pico对应的配置文件。
为了方便用户，截断和配置你们可以完成，然后上传吗？
另外再请问下，有输出onnx的代码吗？如果在Ascend 310B上跑推理的运行时间，我觉得应该要先转换 成onnx后的吧。

**Comment by Stevenzzz1996** [comment]
同求

**Comment by wangguijiepedeval** [comment]
> 非常感谢你的回复，我看了开源的代码，没有FARTrack-Tiny/FARTrack-Nano/FARTrack-Pico对应的配置文件。 为了方便用户，截断和配置你们可以完成，然后上传吗？ 另外再请问下，有输出onnx的代码吗？如果在Ascend 310B上跑推理的运行时间，我觉得应该要先转换 成onnx后的吧。

你好，感谢对FARTrack项目的关注！关于如何加载FARTrack-Tiny/FARTrack-Nano/FARTrack-Pico权重的方法如下：
在lib/models/fartrack_sparse/base_backbone.py下的forward_features函数中，会有如下代码：
```
for j, blk in enumerate(self.blocks): # 0~11
    # if j == 10:
    #     break
    zxs, attn = blk(zxs, return_attention=True, padding_mask=mask)
    attn = attn.detach()
    if attn_sum == None:
        attn_sum = attn
    else:
        attn_sum = attn_sum + attn
for j, blk in enumerate(self.extension): # 0~2
    # if j == 0:
    #     break
    zxs, attn = blk(zxs, return_attention=True, padding_mask=mask)
    attn = attn.detach()
    attn_sum = attn_sum + attn
```
将上下两段注释掉的break逻辑代码分别修改为j == 10和j == 0时对应的就是10层的FARTrack-Nano的权重；类似的上下分别修改为j == 6和j == 0时对应的就是6层的FARTrack-Pico的权重；最后将break逻辑代码全部注释掉，加载的就是正常的15层FARTrack-Tiny权重。

为什么会有这样的设置呢？这是因为我们的自蒸馏方法经过一次性蒸馏后便能够直接使用多个版本，不用对每一个变体进行单独蒸馏和训练。

如果你想要得到能够直接加载使用的FARTrack-Nano和FARTrack-Pico权重可以自行加载15层的原始权重然后截断再重新保存即可。

如果你想在Ascend 310B推理，可以先将模型转换为onnx然后再通过昇腾的 ATC 工具将onnx转换为om文件即可在Ascend 310B上加载om进行推理。

**Comment by canquanshen** [comment]
非常感谢你的回复。
1、自行加载15层的原始权重然后截断再重新保存即可
     --- 这部分有代码吗？
2、如果你想在Ascend 310B推理，可以先将模型转换为onnx然后再通过昇腾的 ATC 工具将onnx转换为om文件即可在Ascend 310B上加载om进行推理。
    --- 请问pytorch 转onnx的代码有吗？

**Comment by canquanshen** [comment]
我自己写了一个pytorch 转onnx的代码，在转换过程中，截断模型。有需要自取。
`
import os
import sys
import torch
from lib.test.evaluation.tracker import Tracker
import onnx
import argparse
import numpy as np
import onnxsim

prj_path = os.path.join(os.path.dirname(__file__), '..')
if prj_path not in sys.path:
    sys.path.append(prj_path)

import importlib

def export():
    parser = argparse.ArgumentParser(description='Run tracker on sequence or dataset.')
    parser.add_argument('tracker_name', type=str, help='Name of tracking method.')
    parser.add_argument('tracker_param', type=str, help='Name of config file.')
    parser.add_argument('onnx_file', type=str, help='output onnx file.')
    parser.add_argument('blk_layer_num', type=int,default=0, help='the number of blk layer.')

    args = parser.parse_args()
    print(args)
    tracker_module = importlib.import_module('lib.test.tracker.{}'.format(args.tracker_name))
    tracker_class = tracker_module.get_tracker_class()
    param_module = importlib.import_module('lib.test.parameter.{}'.format(args.tracker_name))
    params = param_module.parameters(args.tracker_param)
    tracker = tracker_class(params,None)
    model = tracker.network.eval().cpu()
    #截断模型。
    if args.blk_layer_num > 0 and args.blk_layer_num < 15:
        model.backbone.blocks = model.backbone.blocks[:args.blk_layer_num]

    template_imgs = [torch.randn(1,3,112,112).cpu()] * params.cfg.DATA.TEMPLATE.NUMBER
    search_img = torch.randn(1,3,224,224)
    search_img = search_img.cpu()
    template_mask = None
    seqs_out = None
    seqs_out1 = np.array([960,540,60,60])
    seqs_out1 = torch.from_numpy(seqs_out1)
    for i in range(params.cfg.MODEL.PRENUM - 1): # prenum：3
        if i == 0:
            seqs_out = seqs_out1
        else:
            seqs_out = torch.cat((seqs_out, seqs_out1), dim=-1)
    seqs_out = seqs_out.unsqueeze(0)
    seqs_out = seqs_out.cpu()
    stage="inference"
    search_feature=None
    mask=torch.ones([1, 445, 445]) > 0
    mask = mask.cpu()
    input_names = ["template_img0","template_img1","template_img2","template_img3","template_img4","search_img","seqs_out","mask"]
    #output_names = ['seqs',"class","feat","seq_feat","x_feat","mask0","mask1","mask2","mask3"]
    output_names = ['seqs',"class","feat","seq_feat","x_feat","mask_return"]
    
    args1=(template_imgs,search_img,template_mask,None,None,seqs_out,None,stage,search_feature,None,None,mask)
    torch.onnx.export(model=model,args=args1, f=args.onnx_file, input_names=input_names, output_names=output_names,opset_version=12)

    model_onnx = onnx.load(args.onnx_file)  # load onnx model
    onnx.checker.check_model(model_onnx)  # check onnx model
    model_onnx, check = onnxsim.simplify(model_onnx)
    onnx.save(model_onnx, args.onnx_file)

if __name__ == '__main__':
    export()


`

**Comment by AngYang-YA** [comment]
你好，最近也在跑这个代码的测试和训练，方便加一个联系方式一起交流一些问题吗？这时我的qq/qq邮箱：573686178qq.com

**Comment by TsingWei** [comment]
> 我自己写了一个pytorch 转onnx的代码，在转换过程中，截断模型。有需要自取。 

@canquanshen @AlexDotHam 
@wangguijiepedeval 目前的实现只看到用mask来屏蔽token，我在profile时即使构造了75%稀疏的mask喂进去，计算量还是没减。delete token该怎么实现才能对上论文中的计算量？

**Comment by TsingWei** [comment]
> > 我自己写了一个pytorch 转onnx的代码，在转换过程中，截断模型。有需要自取。
> 
> [@canquanshen](https://github.com/canquanshen) [@AlexDotHam](https://github.com/AlexDotHam) [@wangguijiepedeval](https://github.com/wangguijiepedeval) 目前的实现只看到用mask来屏蔽token，我在profile时即使构造了75%稀疏的mask喂进去，计算量还是没减。delete token该怎么实现才能对上论文中的计算量？

我在进入blocks前用mask取模版token，profile的macs基本对上了论文中的数据，但是不确定是否与正确操作有出入。

**Comment by canquanshen** [comment]
> 你好，最近也在跑这个代码的测试和训练，方便加一个联系方式一起交流一些问题吗？这时我的qq/qq邮箱：573686178qq.com

已添加好友，qq很少玩，可以加我微信:13064718889,注明：FARTrack。

**Comment by lianDaniel** [comment]
尝试转onnx，一直在报，对应的算子在onnx上不支持
